cuda/solver/common_trs_kernels.cu

```diff
@@ -0,0 +1,1910 @@
+                    as_cuda_type(b->get_const_values()), b->get_stride(),
+                    as_cuda_type(x->get_values()), x->get_stride(), n, nrhs,
+                    unit_diag, nan_produced.get_data(),
+                    atomic_counter.get_data());
+            } else {
+                sptrsv_naive_legacy_kernel<false><<<grid_size, block_size>>>(
+                    matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                    as_cuda_type(matrix->get_const_values()),
+                    as_cuda_type(b->get_const_values()), b->get_stride(),
+                    as_cuda_type(x->get_values()), x->get_stride(), n, nrhs,
+                    unit_diag, nan_produced.get_data(),
+                    atomic_counter.get_data());
+            }
+        } else {
+            if (is_upper) {
+                sptrsmebrcnm_kernel<true><<<grid_size, block_size>>>(
+                    matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                    as_cuda_type(matrix->get_const_values()),
+                    as_cuda_type(b->get_const_values()), b->get_stride(),
+                    as_cuda_type(x->get_values()), x->get_stride(), n, nrhs,
+                    nan_produced.get_data(), atomic_counter.get_data(), m,
+                    unit_diag);
+            } else {
+                sptrsmebrcnm_kernel<false><<<grid_size, block_size>>>(
+                    matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                    as_cuda_type(matrix->get_const_values()),
+                    as_cuda_type(b->get_const_values()), b->get_stride(),
+                    as_cuda_type(x->get_values()), x->get_stride(), n, nrhs,
+                    nan_produced.get_data(), atomic_counter.get_data(), m,
+                    unit_diag);
+            }
+        }
+#if GKO_VERBOSE_LEVEL >= 1
+        if (exec->copy_val_to_host(nan_produced.get_const_data())) {
+            std::cerr << "Error: triangular solve produced NaN, either not all "
+                         "diagonal "
+                         "elements are nonzero, or the system is very "
+                         "ill-conditioned. "
+                         "The NaN will be replaced with a zero.\n";
+        }
+#endif  // GKO_VERBOSE_LEVEL >= 1
+    }
+};
+
+
+template <typename ValueType, typename IndexType, bool is_upper>
+__global__ void sptrsvelcr_kernel(
+    const IndexType* const rowptrs, const IndexType* const colidxs,
+    const ValueType* const vals, const ValueType* const b,
+    const size_type b_stride, ValueType* const x, const size_type x_stride,
+    const IndexType* const levels, const IndexType sweep, const IndexType n,
+    const IndexType nrhs, bool unit_diag)
+{
+    const auto gid = thread::get_thread_id_flat<IndexType>();
+    const auto row = gid / nrhs;
+    const auto rhs = gid % nrhs;
+
+    if (row >= n) {
+        return;
+    }
+
+    if (levels[row] != sweep) {
+        return;
+    }
```

Comment:
does it make sense to sort by `levels` and instead launch only threads for the rows in that level?